### PR TITLE
fcrypt: tmpdir configuration option

### LINKED ISSUE #1567: Fcrypt: store decrypted temporary files to tmpfs
# Proposal

## Idea

Fcrypt provides another configuration option `fcrypt/tmpdir=<path>`.
The plugin stores the temporarily decrypted GPG output to the specified path.

If no path is provided within the configuration, fcrypt will automatically mount and unmount a `tmpfs` as required.

## Implications

1. The risk of leaking decrypted data is reduced.
2. Thus it would be reasonable to remove the tags: `experimental unfinished discouraged`

## Estimated amount of work

Reasonable.

## Downsides

- Does not work on Windows

## Notes

- An example of mounting `tmpfs` can be found [here](https://gist.github.com/gcmurphy/c4c5222075d8e501d7d1).
- Mounting `tmpfs` is required at `kdb get` / `pregetstorage`
- Unmounting `tmpfs` is required at `kdb get` / `postgetstorage`

### PR DESCRIPTION
**Pull request by petermax2**
# Purpose

Introduce the configuration option `fcrypt/tmpdir` to `fcrypt`.
`fcrypt` uses the given path to redirect GPG output to.
The default value is `/tmp`.

Closes #1567 .

# Checklist

Please only check relevant points.
For docu fixes, spell checking and similar nothing
needs to be checked.

- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went fine
- [ ] I added unit tests
- [x] affected documentation is fixed
- [ ] I added code comments, logging, and assertions
- [ ] meta data is updated (e.g. README.md of plugins)

@markus2330 please review my pull request

### CODE REVIEW

**Comment by petermax2** [comment]
@markus2330 as discussed I implemented the `tmpdir` configuration option for `fcrypt`. Please have a look at `contract.h`. Is this intended use of `config/needs` or should we remove this line from the contract?

**Comment by petermax2** [comment]
**TODO**

- [ ] remove `config/needs` from `contract.h` 
- [ ] set default value for `tmpdir` in `kdbOpen()` or maybe `checkconf()`

File: src/plugins/fcrypt/README.md [hunk 1]
```diff
@@ -23,8 +23,9 @@ After the getstorage plugin has read the backend file, the plugin decrypts the b
 
 ## Security Considerations
 
-During decryption the plugin temporarily writes the decrypted plain text to the same directory as the original (encrypted) file.
-This is a vulnerability as an attacker might have access to the plain text for a short period of time (the time between pregetstorage and postgetstorage calls).
+During decryption the plugin temporarily writes the decrypted plain text to a configurable temporary directory (`/tmp` is used as default).
+This might be a vulnerability as an attacker might have access to the plain text for a short period of time (the time between pregetstorage and postgetstorage calls).
```
Comment (markus2330) [inline src/plugins/fcrypt/README.md]: Its not clear for me from this text that there are actually two problems:

1. A file is shortly visible in a filesystem. Protection: use a directly which is only readable by the user. Is this already addressed in this PR?
2. Bytes of private data might end up on the hard disc. Protection: tmpfs

File: src/plugins/fcrypt/README.md [hunk 2]
```diff
@@ -119,3 +120,12 @@ output of `fcrypt` is ASCII armored. If no encryption key is provided (i.e. only
...
 Textmode can be disabled by setting `fcrypt/textmode` to `0` in the plugin configuration.
 
+### Temporary Directory
+
+`fcrypt` uses the configuration option `fcrypt/tmpdir` to generate paths for temporary files during encryption and decryption.
+The path is forwarded to GPG via the `-o` option, so GPG will output to this path.
+
+`/tmp` is used as default value.
+
+We recommend to specify a path that is mounted to a RAM disk.
```
Comment (markus2330) [inline src/plugins/fcrypt/README.md]: You should mention that it needs to be read/writeable by user.

File: src/plugins/fcrypt/contract.h [hunk 3]
```diff
@@ -20,3 +20,7 @@ keyNew ("system/elektra/modules/" ELEKTRA_PLUGIN_NAME, KEY_VALUE, "fcrypt plugin
 		ELEKTRA_PLUGIN_FUNCTION (ELEKTRA_PLUGIN_NAME, checkconf), KEY_END),
 #include ELEKTRA_README (fcrypt)
 	keyNew ("system/elektra/modules/" ELEKTRA_PLUGIN_NAME "/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END),
+	keyNew ("system/elektra/modules/" ELEKTRA_PLUGIN_NAME "/constants/DEFAULT_TMPDIR", KEY_VALUE, ELEKTRA_FCRYPT_DEFAULT_TMPDIR,
+		KEY_END),
+	keyNew ("system/elektra/modules/" ELEKTRA_PLUGIN_NAME "/config/needs" ELEKTRA_FCRYPT_CONFIG_TMPDIR, KEY_VALUE,
```
Comment (markus2330) [inline src/plugins/fcrypt/contract.h]: Ahh nice, doing so you write the tmpdir into the system namespace of plugin config, but the user can overwrite it during mount (which will be placed into the user namespace of the plugin config).

File: src/plugins/fcrypt/fcrypt.c [hunk 4]
```diff
@@ -17,6 +17,7 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <gpg.h>
+#include <libgen.h>
```
Comment (markus2330) [inline src/plugins/fcrypt/fcrypt.c]: Can you write in a comment the purpose of this include file? (basename?)

File: src/plugins/fcrypt/fcrypt.c [hunk 5]
```diff
@@ -48,19 +49,47 @@ typedef struct _fcryptState fcryptState;
...
+ * @param conf holds the plugin configuration
  * @param file holds the path to the original file
  * @param fd will hold the file descriptor to the temporary file in case of success
  * @returns an allocated string holding the name of the encrypted file. Must be freed by the caller.
  */
-static char * getTemporaryFileName (const char * file, int * fd)
+static char * getTemporaryFileName (KeySet * conf, const char * file, int * fd)
 {
+	// read the temporary directory to use from the plugin configuration
+	const char * tmpDir;
```
Comment (markus2330) [inline src/plugins/fcrypt/fcrypt.c]: `= ELEKTRA_FCRYPT_DEFAULT_TMPDIR` and you do not need `else` below.

File: src/plugins/fcrypt/fcrypt.c [hunk 6]
```diff
@@ -48,19 +49,47 @@ typedef struct _fcryptState fcryptState;
...
-	snprintf (newFile, newFileAllocated, "%s" ELEKTRA_FCRYPT_TMP_FILE_SUFFIX, file);
+	if (!newFile) goto error;
+	snprintf (newFile, newFileAllocated, "%s/%s" ELEKTRA_FCRYPT_TMP_FILE_SUFFIX, tmpDir, baseName);
 	*fd = mkstemp (newFile);
+
+	elektraFree (fileDup);
 	return newFile;
+
+error:
+	if (fileDup) elektraFree (fileDup);
```
Comment (markus2330) [inline src/plugins/fcrypt/fcrypt.c]: It is safe to free a nullptr.

File: src/plugins/fcrypt/fcrypt.c [hunk 7]
```diff
@@ -48,19 +49,47 @@ typedef struct _fcryptState fcryptState;
...
-	const size_t newFileAllocated = strlen (file) + strlen (ELEKTRA_FCRYPT_TMP_FILE_SUFFIX) + 1;
+	// ----------------------------------------------
+	// + 2 characters in total
+	const size_t newFileAllocated = strlen (tmpDir) + strlen (baseName) + strlen (ELEKTRA_FCRYPT_TMP_FILE_SUFFIX) + 2;
 	char * newFile = elektraMalloc (newFileAllocated);
-	if (!newFile) return NULL;
-	snprintf (newFile, newFileAllocated, "%s" ELEKTRA_FCRYPT_TMP_FILE_SUFFIX, file);
+	if (!newFile) goto error;
+	snprintf (newFile, newFileAllocated, "%s/%s" ELEKTRA_FCRYPT_TMP_FILE_SUFFIX, tmpDir, baseName);
 	*fd = mkstemp (newFile);
```
Comment (markus2330) [inline src/plugins/fcrypt/fcrypt.c]: mkstemp might fail (return -1)

**Comment by petermax2** [comment]
@markus2330 thank you for the feedback. The documentation should be easier to understand now.

**Comment by markus2330** [comment]
Great to see the update! Thank you!

**Pull request merged**